File: Python/2294_동전2.py
if __name__ == '__main__':
	n,k = list(map(int,input().split()))
	
	coin = [] # 동전 종류를 담을 배열
	for _ in range(n):
		coin.append(input())
		
	coin = list(map(int, coin))
	
	dp = [float('inf')]*(k+1) # 사용할 동전의 최솟값을 담을 dp 배열 (무한대로 세팅해놓기), dp[i]는 가치 i를 만들 때 사용되는 최소 동전의 개수를 의미
	dp[0] = 0 # dp[0]은 0으로 초기화
	for c in coin:
		for i in range(c,k+1): # 동전의 가치보다 커야 동전을 사용할 수 있음
			dp[i] = min(dp[i],1+dp[i-c]) # 현재 동전을 1개 사용할 수 있고, 현재 가치에서 현재 동전 가치를 뺀만큼을 추가로 사용해야 함
               # 예를 들어 현재 만들고자 하는 가치가 8이고, 현재 동전의 가치가 5라면, 현재 동전 1개를 사용하고 8-5, 즉, 3만큼을 추가로 만들어야 함(이는 dp[3]의 값을 가져오면 됨)
				
	if dp[k] == float('inf'): # 배열 값이 무한대라면, 만들 수 없음을 의미
		print(-1) # -1 출력
	else: # 만들 수 있다면
		print(dp[k]) # dp[k]를 출력

File: C++/2294_동전2.cpp
#include <iostream>
#include <algorithm>
using namespace std;

int main() {
	int n, k;
	cin >> n >> k;
	
	int dp[k+1];
	for(int i = 0; i <= k; i++){
		dp[i] = 987654321; // 큰 값으로 초기화
	}
	
	dp[0] = 0; // dp[0]은 0으로 초기화
	for(int c = 0; c < n; c++){
		int coin;
		cin >> coin;
		
		for(int i = coin; i <= k; i++){ // 구하고자하는 가치가 현재 동전 가치보다 커야 동전을 사용할 수 있음
			dp[i] = min(dp[i],1+dp[i-coin]); // 현재 가치의 동전을 사용한 경우와 그렇지 않은 경우 중 더 작은 값을 dp에 저장
		}
	}
	
	if(dp[k] == 987654321){ // 만약 dp의 값이 갱신되지 않고 큰 값으로 남아 있다면, 현재 가지고 있는 동전으로 원하는 가치를 만들 수 없다는 의미
		dp[k] = -1; // dp[k]를 -1로 갱신
	}
	
	cout << dp[k];
	
	return 0;
}